frontend/src/components/widgets/Radio/Radio.tsx

```diff
@@ -14,128 +14,197 @@
-    this.formClearHelper.disconnect()
+  const updateFromProtobuf = (): void => {
+    element.setValue = false
+    setStateValue(element.value)
+    setStateSource({ fromUi: false })
   }
 
-  private maybeUpdateFromProtobuf(): void {
-    const { setValue } = this.props.element
+  const maybeUpdateFromProtobuf = (): void => {
+    const { setValue } = element
     if (setValue) {
-      this.updateFromProtobuf()
+      updateFromProtobuf()
     }
   }
 
-  private updateFromProtobuf(): void {
-    const { value } = this.props.element
-    this.props.element.setValue = false
-    this.setState({ value }, () => {
-      this.commitWidgetValue({ fromUi: false })
-    })
+  const checkNoOptions = (): void => {
+    if (options.length === 0) {
+      options.push("No options to select.")
+    }
+    if (options[0] === "No options to select.") {
+      disabled = true
+    }
   }
 
-  /** Commit state.value to the WidgetStateManager. */
-  private commitWidgetValue = (source: Source): void => {
-    this.props.widgetMgr.setIntValue(
-      this.props.element,
-      this.state.value,
-      source
-    )
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    const selectedIndex = parseInt(e.target.value, 10)
+    setStateValue(selectedIndex)
+    setStateSource({ fromUi: true })
   }
 
   /**
    * If we're part of a clear_on_submit form, this will be called when our
    * form is submitted. Restore our default value and update the WidgetManager.
    */
-  private onFormCleared = (): void => {
-    this.setState(
-      (_, prevProps) => {
-        return { value: prevProps.element.default }
-      },
-      () => this.commitWidgetValue({ fromUi: true })
-    )
+  const onFormCleared = (): void => {
+    setStateValue(element.default)
+    setStateSource({ fromUi: true })
   }
 
-  private onChange = (selectedIndex: number): void => {
-    this.setState({ value: selectedIndex }, () =>
-      this.commitWidgetValue({ fromUi: true })
+  useEffect(() => {
+    if (element.setValue) {
+      updateFromProtobuf()
+    } else {
+      setStateSource({ fromUi: false })
+    }
+    // Add form-clear event handler
+    const formListener = widgetMgr.addFormClearedListener(
+      element.formId,
+      onFormCleared
     )
-  }
+    // Remove form-clear listener when component unmounts
+    return function cleanup() {
+      formListener.disconnect()
+    }
+  }, [])
 
-  public render(): React.ReactNode {
-    const { disabled, element, width, widgetMgr } = this.props
-    const { horizontal, options, label, labelVisibility, help } = element
+  // Commit state value/source to the WidgetStateManager.
+  useEffect(() => {
+    props.widgetMgr.setIntValue(element, value, source)
+  }, [value, source])
 
-    // Manage our form-clear event handler.
-    this.formClearHelper.manageFormClearListener(
-      widgetMgr,
-      element.formId,
-      this.onFormCleared
-    )
+  useEffect(() => {
+    maybeUpdateFromProtobuf()
+  })
```

Comment:
Hm, does this need to be called on every component render? I think we can add `element` to this dependency array so that this doesn't get called unless it has to be (even if the function is a pretty cheap one to execute)